fix(sys): load exclusively using dynamic linking

Beanow opened the pull request:
<!--
Please make sure to read the Pull Request Guidelines:
https://github.com/tauri-apps/tauri/blob/dev/.github/CONTRIBUTING.md#pull-request-guidelines
-->

<!-- PULL REQUEST TEMPLATE -->
<!-- (Update "[ ]" to "[x]" to check a box) -->

**What kind of change does this PR introduce?** (check at least one)

- [x] Bugfix
- [ ] Feature
- [ ] Code style update
- [x] Refactor
- [ ] Documentation
- [ ] Build-related changes
- [ ] Other, please describe:

**Does this PR introduce a breaking change?** (check one)
<!--
If yes, please describe the impact and migration path for existing applications in an attached issue. Filing a PR with breaking changes that has not been discussed and approved by the maintainers in an issue will be immediately closed.
-->

- [x] Yes. Issue #31
- [ ] No

It would remove support for `$APPDIR` based library paths applied only inside an appimage.

**The PR fulfills these requirements:**

- [ ] When resolving a specific issue, it's referenced in the PR's title (e.g. `fix: #xxx[,#xxx]`, where "xxx" is the issue number)
- [x] A change file is added if any packages will require a version bump due to this PR per [the instructions in the readme](https://github.com/tauri-apps/tauri/blob/dev/.changes/readme.md).

If adding a **new feature**, the PR's description includes:
- [ ] A convincing reason for adding this feature (to avoid wasting your time, it's best to open a suggestion issue first and wait for approval before working on it)

**Other information:**

This should let `dlopen` (through `ld.so`) handle what paths to search in for the respective libraries.

Additionally this fixes a mistake with the library filenames. Now using the `SONAME` instead of a symlinked name that happened to work when dev packages are installed.

Some major benefits:
- No more hard-coded paths, so other environments (like flatpak) can be supported.
- Removes `pkg-config` dependency, which is considered a _build tool_ and usually isn't available on a clean install.
- `LD_LIBRARY_PATH` can be used if all else fails.

lucasfernog commented:
I can test this tomorrow.

lucasfernog commented on .changes/fix-dynamic-linking.md:
```diff
@@ -0,0 +1,12 @@
+
+---
+"libappindicator": major
```
the bump should be `minor` instead, we're not changing this crate to stable yet.

lucasfernog commented:
Ahh I just checked the discussion, this PR should be ok once we can fix the appimage bundler. Though I wouldn't merge it before that one is tested.

Beanow commented:
@lucasfernog so I'm thinking we need some kind of "backcompat" feature flag here.

`v0.7.1` and `v0.7.2` _essentially_ implemented a 'risky' dependency, by loading `libappindicator3.so`.

Interpreting this by shared library naming conventions,
This 'means': load the _latest_ version available on the system. And **don't care about ABI compatibility**.

Instead it should be: `libappindicator3.so.1`. Meaning: load a version **v1 ABI compatible** version.
And typically, distros will symlink this to the complete semver-like version. E.g. `libappindicator3.so.1.0.0`

---

Because, if someday ayatana decides to release an ABI breaking change, our crate fail if we load a symlink set up like `libayatana-appindicator3.so -> libayatana-appindicator3.so.2.0.0`.
So the correct thing to do would be requesting `libayatana-appindicator3.so.1` specifically.

This allows the distro to install both versions at the same time, something like:

```
libayatana-appindicator3.so -> libayatana-appindicator3.so.2.0.0
libayatana-appindicator3.so.1 -> libayatana-appindicator3.so.1.0.0
libayatana-appindicator3.so.1.0.0
libayatana-appindicator3.so.2 -> libayatana-appindicator3.so.2.0.0
libayatana-appindicator3.so.2.0.0
```

---

Why we'd need backcompat, is because previous versions relied on two things:

1. A `libayatana-appindicator3.so` or `libappindicator3.so` file exists at all.
   (This is not true for Ubuntu's `libappindicator3-1` or `libayatana-appindicator3-1` packages, only the `-dev`)
2. Those files are v1 ABI compatible.
   (This has been true so far, because I've only ever seen `v1.0.0` out in the wild)

However looking at Tauri's appimage bundler, we **only** provide `libayatana-appindicator3.so` or `libappindicator3.so`. Without any symlinks set up. So if we start to load `.1`,  we would break older appimage bundlers (CLI) versions.

So we could use a `v0.7-backcompat` feature, to also look for the old filenames. To give people a grace period to upgrade their Tauri CLI's.

Beanow commented:
Haha so something interesting happening, I'm testing with (unchanged) appimage bundles:

Because our search order is now:
1. `libayatana-appindicator3.so.1`
1. `libappindicator3.so.1`
1. `libayatana-appindicator3.so`
1. `libappindicator3.so`

My system matches `/usr/lib/libappindicator3.so.1` the *system library*, because the appimage bundles `libappindicator3.so` which is the last one it tries to find.

I'll need to test the backcompat from a docker container I guess.

```
     32360:	find library=libayatana-appindicator3.so.1 [0]; searching
     32360:	 search path=/tmp/.mount_tauri-RhXf2M/usr/lib:/tmp/.mount_tauri-RhXf2M/usr/lib32:glibc-hwcaps/x86-64-v3:glibc-hwcaps/x86-64-v2:tls/x86_64/x86_64:tls/x86_64:tls/x86_64:tls:x86_64/x86_64:x86_64:x86_64:		(LD_LIBRARY_PATH)
     32360:	  trying file=/tmp/.mount_tauri-RhXf2M/usr/lib/libayatana-appindicator3.so.1
     32360:	  trying file=/tmp/.mount_tauri-RhXf2M/usr/lib32/libayatana-appindicator3.so.1
     32360:	  trying file=glibc-hwcaps/x86-64-v3/libayatana-appindicator3.so.1
     32360:	  trying file=glibc-hwcaps/x86-64-v2/libayatana-appindicator3.so.1
     32360:	  trying file=tls/x86_64/x86_64/libayatana-appindicator3.so.1
     32360:	  trying file=tls/x86_64/libayatana-appindicator3.so.1
     32360:	  trying file=tls/x86_64/libayatana-appindicator3.so.1
     32360:	  trying file=tls/libayatana-appindicator3.so.1
     32360:	  trying file=x86_64/x86_64/libayatana-appindicator3.so.1
     32360:	  trying file=x86_64/libayatana-appindicator3.so.1
     32360:	  trying file=x86_64/libayatana-appindicator3.so.1
     32360:	  trying file=libayatana-appindicator3.so.1
     32360:	 search path=/tmp/.mount_tauri-RhXf2M/usr/bin/../lib		(RUNPATH from file tauri-api)
     32360:	  trying file=/tmp/.mount_tauri-RhXf2M/usr/bin/../lib/libayatana-appindicator3.so.1
     32360:	 search cache=/etc/ld.so.cache
     32360:	 search path=/usr/lib/glibc-hwcaps/x86-64-v3:/usr/lib/glibc-hwcaps/x86-64-v2:/usr/lib/tls/x86_64/x86_64:/usr/lib/tls/x86_64:/usr/lib/tls/x86_64:/usr/lib/tls:/usr/lib/x86_64/x86_64:/usr/lib/x86_64:/usr/lib/x86_64:/usr/lib		(system search path)
     32360:	  trying file=/usr/lib/glibc-hwcaps/x86-64-v3/libayatana-appindicator3.so.1
     32360:	  trying file=/usr/lib/glibc-hwcaps/x86-64-v2/libayatana-appindicator3.so.1
     32360:	  trying file=/usr/lib/tls/x86_64/x86_64/libayatana-appindicator3.so.1
     32360:	  trying file=/usr/lib/tls/x86_64/libayatana-appindicator3.so.1
     32360:	  trying file=/usr/lib/tls/x86_64/libayatana-appindicator3.so.1
     32360:	  trying file=/usr/lib/tls/libayatana-appindicator3.so.1
     32360:	  trying file=/usr/lib/x86_64/x86_64/libayatana-appindicator3.so.1
     32360:	  trying file=/usr/lib/x86_64/libayatana-appindicator3.so.1
     32360:	  trying file=/usr/lib/x86_64/libayatana-appindicator3.so.1
     32360:	  trying file=/usr/lib/libayatana-appindicator3.so.1
     32360:	
     32360:	find library=libappindicator3.so.1 [0]; searching
     32360:	 search path=/tmp/.mount_tauri-RhXf2M/usr/lib:/tmp/.mount_tauri-RhXf2M/usr/lib32:glibc-hwcaps/x86-64-v3:glibc-hwcaps/x86-64-v2:tls/x86_64/x86_64:tls/x86_64:tls/x86_64:tls:x86_64/x86_64:x86_64:x86_64:		(LD_LIBRARY_PATH)
     32360:	  trying file=/tmp/.mount_tauri-RhXf2M/usr/lib/libappindicator3.so.1
     32360:	  trying file=/tmp/.mount_tauri-RhXf2M/usr/lib32/libappindicator3.so.1
     32360:	  trying file=glibc-hwcaps/x86-64-v3/libappindicator3.so.1
     32360:	  trying file=glibc-hwcaps/x86-64-v2/libappindicator3.so.1
     32360:	  trying file=tls/x86_64/x86_64/libappindicator3.so.1
     32360:	  trying file=tls/x86_64/libappindicator3.so.1
     32360:	  trying file=tls/x86_64/libappindicator3.so.1
     32360:	  trying file=tls/libappindicator3.so.1
     32360:	  trying file=x86_64/x86_64/libappindicator3.so.1
     32360:	  trying file=x86_64/libappindicator3.so.1
     32360:	  trying file=x86_64/libappindicator3.so.1
     32360:	  trying file=libappindicator3.so.1
     32360:	 search path=/tmp/.mount_tauri-RhXf2M/usr/bin/../lib		(RUNPATH from file tauri-api)
     32360:	  trying file=/tmp/.mount_tauri-RhXf2M/usr/bin/../lib/libappindicator3.so.1
     32360:	 search cache=/etc/ld.so.cache
     32360:	  trying file=/usr/lib/libappindicator3.so.1
```

Beanow commented:
So the TL;DR this PR does not break appimage bundles with the feature flag enabled. Though it may start using your system library instead of the bundled one. But it would be good hygiene to fix the bundler anyway so it provides the `.1` suffixed files.

lucasfernog commented:
Why do we need a compat feature if we're doing a minor bump so tao/wry will need to update anyway?

Beanow commented:
@lucasfernog The compatibility is because the Tauri CLI (thus bundler) you have installed may update independently from your Cargo dependencies.

I've come up with a way to add a compatibility fix to the bundler as well, and it would allow both ways to _just work_ when updated independently.

Testing with that:

- fixed bundler + fixed libappindicator-sys: It prefers the bundled `/tmp/.mount_tauri-uy8bqq/usr/lib/libappindicator3.so.1` over system lib.
- fixed bundler + hack brach libappindicator-sys (https://github.com/Beanow/libappindicator-rs/commit/2aa4263400431b355a80deab336c91401e5198af): It prefers the bundled **symlink** `/tmp/.mount_tauri-N0Hons/usr/lib/libappindicator3.so` over system libs.
- fixed bundler +  libappindicator-sys v0.7.2 release: Does not dynamically search with `dlopen` and opens the bundled **symlink** `/tmp/.mount_tauri-HLUzNY/usr/lib/libappindicator3.so`
 - current bundler + fixed libappindicator-sys: It chose my system `/usr/lib/libappindicator3.so.1` because it happens to be installed (may be unexpected by users)
- current bundler + hack brach libappindicator-sys: It uses bundled `/tmp/.mount_tauri-R4D5hk/usr/lib/libappindicator3.so`.
- current bundler +  libappindicator-sys v0.7.2 release: Does not dynamically search with dlopen and opens the bundled **regular file** `/tmp/.mount_tauri-lDeIBc/usr/lib/libappindicator3.so`
